src/lib/Err.hs

```diff
@@ -42,61 +54,142 @@ data ErrType = NoErr
              | ZipErr
              | EscapedNameErr
              | ModuleImportErr
+             | MonadFailErr
                deriving (Show, Eq)
 
-type Except = Either Err
+type SrcPosCtx  = Maybe SrcPos
+type SrcTextCtx = Maybe (Int, String) -- Int is the offset in the source file
+data ErrCtx = ErrCtx
+  { srcTextCtx :: SrcTextCtx
+  , srcPosCtx  :: SrcPosCtx
+  , messageCtx :: [String] }
+    deriving (Show, Eq)
+
+type Except = Either Errs
 
 type SrcPos = (Int, Int)
-type SrcCtx = Maybe SrcPos
 
-type MonadErr = MonadError Err
+class MonadFail m => Fallible m where
+  throwErrs :: Errs -> m a
+  addErrCtx :: ErrCtx -> m a -> m a
+
+-- We have this in its own class because IO and `Either Err` can't implement it
+class Fallible m => CtxReader m where
+  getErrCtx :: m ErrCtx
+
+-- We have this in its own class because StateT can't implement it
+class Fallible m => FallibleApplicative m where
+  mergeErrs :: m a -> m b -> m (a, b)
+
+newtype FallibleM a =
+  FallibleM { fromFallibleM :: ReaderT ErrCtx (Either Errs) a }
+  deriving (Functor, Applicative, Monad)
+
+instance Fallible FallibleM where
+  throwErrs errs = FallibleM $ lift $ throwErrs errs
+  addErrCtx ctx (FallibleM m) = FallibleM $ local (<> ctx) m
+
+instance FallibleApplicative FallibleM where
+  mergeErrs (FallibleM (ReaderT f1)) (FallibleM (ReaderT f2)) =
+    FallibleM $ ReaderT \ctx -> mergeErrs (f1 ctx) (f2 ctx)
+
+instance CtxReader FallibleM where
+  getErrCtx = FallibleM ask
+
+instance Fallible IO where
+  throwErrs errs = throwIO errs
+  addErrCtx ctx m = do
+    result <- catchIOExcept m
+    liftExcept $ addErrCtx ctx result
+
+instance FallibleApplicative IO where
+  mergeErrs m1 m2 = do
+    result1 <- catchIOExcept m1
+    result2 <- catchIOExcept m2
+    liftExcept $ mergeErrs result1 result2
+
+-- === SeqFallibleApplicative ===
+
+-- Wraps a Fallible monad, presenting an applicative interface that sequences
+-- actions using the error-concatenating `mergeErrs` instead of the default
+-- abort-on-failure sequencing.
```

Comment:
I'm not sure if I fully understand the difference between `FallibleApplicative` and `SeqFaillableApplicative`. Could you please elaborate why we need both?

Reply:
We might not need `SeqFallibleApplicative` at all (we don't use it currently) but it's different from `FallibleApplicative` because it's a data type, not a class. `SeqFallibleApplicative` is a wrapper that takes a type that implements `FallibleApplicative` and uses its `mergeErrors` method to implement `Applicative`. The type being wrapped probably implements `Applicative` too, but in the usual abort-on-failure way in order to be consistent with its monad instance. Does that make sense?

Reply:
Ah yes, that makes total sense. Thanks for explaining. The name is a little confusing, but tbh I don't have better ideas so we can keep it for now.

Reply:
Agreed, it's a bad name. Maybe I'll try something that at least has `Wrapper` in the name.